File: lib/features/journey/models/activity_insights.dart
class StrongestActivityDay {
  const StrongestActivityDay({
    required this.date,
    required this.dailyScore,
    required this.steps,
  });

  factory StrongestActivityDay.fromJson(Map<String, dynamic> json) {
    return StrongestActivityDay(
      date: DateTime.parse(json['date'] as String),
      dailyScore: (json['daily_score'] as num).toDouble(),
      steps: (json['steps'] as num).toDouble(),
    );
  }

  final DateTime date;
  final double dailyScore;
  final double steps;
}

class ActivityInsights {
  const ActivityInsights({
    required this.days,
    required this.currentRecordedDays,
    required this.previousRecordedDays,
    this.currentLegacyDays = 0,
    this.previousLegacyDays = 0,
    this.totalSteps = 0,
    this.averageSteps,
    this.totalDistance = 0,
    this.totalActiveCalories = 0,
    this.averageScore,
    required this.stepsChangePercent,
    this.distanceChangePercent,
    this.activeCaloriesChangePercent,
    required this.averageScoreChange,
    this.trend = 'insufficient_data',
    this.consistencyDays = 0,
    this.strongestStepsDay,
    StrongestActivityDay? strongestScoreDay,
    StrongestActivityDay? strongestDay,
  }) : strongestScoreDay = strongestScoreDay ?? strongestDay;

  factory ActivityInsights.fromJson(Map<String, dynamic> json) {
    final strongestSteps = json['strongest_steps_day'];
    final strongestScore = json['strongest_score_day'];
    return ActivityInsights(
      days: json['days'] as int,
      currentRecordedDays: json['current_recorded_days'] as int,
      previousRecordedDays: json['previous_recorded_days'] as int,
      currentLegacyDays: json['current_legacy_days'] as int,
      previousLegacyDays: json['previous_legacy_days'] as int,
      totalSteps: (json['total_steps'] as num).toDouble(),
      averageSteps: (json['average_steps'] as num?)?.toDouble(),
      totalDistance: (json['total_distance'] as num).toDouble(),
      totalActiveCalories: (json['total_active_calories'] as num).toDouble(),
      averageScore: (json['average_score'] as num?)?.toDouble(),
      stepsChangePercent: (json['steps_change_percent'] as num?)?.toDouble(),
      distanceChangePercent: (json['distance_change_percent'] as num?)
          ?.toDouble(),
      activeCaloriesChangePercent:
          (json['active_calories_change_percent'] as num?)?.toDouble(),
      averageScoreChange: (json['average_score_change'] as num?)?.toDouble(),
      trend: json['trend'] as String? ?? 'insufficient_data',
      consistencyDays: json['consistency_days'] as int? ?? 0,
      strongestStepsDay: strongestSteps is Map<String, dynamic>
          ? StrongestActivityDay.fromJson(strongestSteps)
          : null,
      strongestScoreDay: strongestScore is Map<String, dynamic>
          ? StrongestActivityDay.fromJson(strongestScore)
          : null,
    );
  }

  final int days;
  final int currentRecordedDays;
  final int previousRecordedDays;
  final int currentLegacyDays;
  final int previousLegacyDays;
  final double totalSteps;
  final double? averageSteps;
  final double totalDistance;
  final double totalActiveCalories;
  final double? averageScore;
  final double? stepsChangePercent;
  final double? distanceChangePercent;
  final double? activeCaloriesChangePercent;
  final double? averageScoreChange;
  final String trend;
  final int consistencyDays;
  final StrongestActivityDay? strongestStepsDay;
  final StrongestActivityDay? strongestScoreDay;

  StrongestActivityDay? get strongestDay => strongestScoreDay;

  bool get hasComparablePeriods =>
      currentRecordedDays > 0 && previousRecordedDays > 0;
}

File: backend/app/services/activity_insights.py
from collections.abc import Sequence
from statistics import fmean

from app.db.models.activity import Activity

_TREND_THRESHOLD = 5.0


def determine_trend(
    current_recorded_days: int,
    previous_recorded_days: int,
    average_score_change: float | None,
) -> str:
    """Return a deterministic trend label from confirmed recorded data only."""
    if (
        current_recorded_days == 0
        or previous_recorded_days == 0
        or average_score_change is None
    ):
        return "insufficient_data"
    if average_score_change > _TREND_THRESHOLD:
        return "improving"
    if average_score_change < -_TREND_THRESHOLD:
        return "declining"
    return "stable"


def _average(records: Sequence[Activity], field: str) -> float | None:
    vals = [float(getattr(record, field)) for record in records if getattr(record, field) is not None]
    if not vals:
        return None
    return float(fmean(vals))


def _percent_change(current: float | None, previous: float | None) -> float | None:
    if current is None or previous is None or previous == 0:
        return None
    return ((current - previous) / previous) * 100


def build_activity_insights(
    current: Sequence[Activity],
    previous: Sequence[Activity],
    *,
    current_legacy_days: int = 0,
    previous_legacy_days: int = 0,
) -> dict[str, object]:
    current_steps = _average(current, "steps")
    previous_steps = _average(previous, "steps")
    current_score = _average(current, "daily_score")
    previous_score = _average(previous, "daily_score")
    average_score_change: float | None = (
        current_score - previous_score
        if current_score is not None and previous_score is not None
        else None
    )
    strongest_steps = (
        max(current, key=lambda item: (item.steps or 0.0, item.daily_score or 0.0, item.date))
        if current
        else None
    )
    strongest_score = (
        max(current, key=lambda item: (item.daily_score or 0.0, item.steps or 0.0, item.date))
        if current
        else None
    )
    trend = determine_trend(len(current), len(previous), average_score_change)
    return {
        "current_recorded_days": len(current),
        "previous_recorded_days": len(previous),
        "current_legacy_days": current_legacy_days,
        "previous_legacy_days": previous_legacy_days,
        "total_steps": sum(float(item.steps or 0.0) for item in current),
        "average_steps": current_steps,
        "total_distance": sum(float(item.distance or 0.0) for item in current),
        "total_active_calories": sum(float(item.calories or 0.0) for item in current),
        "average_score": current_score,
        "steps_change_percent": _percent_change(current_steps, previous_steps),
        "distance_change_percent": _percent_change(
            sum(float(item.distance) for item in current if item.distance is not None) if current else None,
            sum(float(item.distance) for item in previous if item.distance is not None) if previous else None,
        ),
        "active_calories_change_percent": _percent_change(
            sum(float(item.calories) for item in current if item.calories is not None) if current else None,
            sum(float(item.calories) for item in previous if item.calories is not None) if previous else None,
        ),
        "average_score_change": average_score_change,
        "trend": trend,
        "consistency_days": len(current),
        "strongest_steps_day": strongest_steps,
        "strongest_score_day": strongest_score,
    }
